scheduler integration benchmark improvements

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

I was running the `test/integration/scheduler_perf` and encountered various issues in it and some of the scheduler code. See the individual commit messages for details.

#### Special notes for your reviewer:

I still have problems running all benchmarks in a single "go test" invocation as discussed in https://kubernetes.slack.com/archives/C09TP78DV/p1642623117030500 but this PR helps a bit.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

pohly committed:
kube-scheduler: downgrade namespace log message from "error" to "info"

GetNamespaceLabelsSnapshot has a fallback when it gets errors when looking up a
namespace, therefore reporting the error is more informational than a real
error. In particular, not finding the namespace is normal when running
test/integration/scheduler_perf and happens so frequently that there is a lot
of output on stderr:

E0120 12:19:09.204768   95305 plugin.go:138] "getting namespace, assuming empty set of namespace labels" err="namespace \"namespace-1\" not found" namespace="namespace-1"

pohly committed:
test: allow controlling etcd log level

When running an integration test that measures performance, like for example
test/integration/scheduler_perf, running etcd with debug level output is
undesirable because it creates additional load on the system and isn't
realistic.

The default is still "debug", but ETCD_LOGLEVEL=warn can be used to override
that.

pohly committed:
test: fix info message for etcd startup

The output redirection in the info message did not match the actual
invocation (stdout and stderr swapped).

pohly committed:
scheduler_perf: avoid ambiguous test names

"-bench=PerfScheduling/Preemption/500Nodes" ran both the
PerfScheduling/Preemption/500Nodes and the
PerfScheduling/PreemptionPVs/500Nodes benchmark.

This can be avoided by choosing names where none is the prefix of another.

pohly committed:
scheduler_perf: dump test data when writing it failed

Occasionally, writing as JSON failed because a NaN float couldn't be
encoded. The extended log message helps understand where that comes from, for
example:

F0120 20:24:45.515745  511835 scheduler_perf_test.go:540] BenchmarkPerfScheduling: unable to write measured data {Version:v1 DataItems:[{Data:map[Average:35.714285714285715 Perc50:2 Perc90:36 Perc95:412 Perc99:412] Unit:pods/s Labels:map[Metric:SchedulingThroughput Name:BenchmarkPerfScheduling/PreemptionPVs/500Nodes/namespace-2]} {Data:map[Average:27.863967530999993 Perc50:13.925925925925926 Perc90:30.06711409395973 Perc95:31.85682326621924 Perc99:704] Unit:ms Labels:map[Metric:scheduler_e2e_scheduling_duration_seconds Name:BenchmarkPerfScheduling/PreemptionPVs/500Nodes/namespace-2]} {Data:map[Average:11915.651577744 Perc50:15168.796680497926 Perc90:19417.759336099585 Perc95:19948.87966804979 Perc99:20373.77593360996] Unit:ms Labels:map[Metric:scheduler_pod_scheduling_duration_seconds Name:BenchmarkPerfScheduling/PreemptionPVs/500Nodes/namespace-2]} {Data:map[Average:1.1865832049999983 Perc50:0.7636363636363637 Perc90:2.891903719912473 Perc95:3.066958424507659 Perc99:5.333333333333334] Unit:ms Labels:map[Metric:scheduler_framework_extension_point_duration_seconds Name:BenchmarkPerfScheduling/PreemptionPVs/500Nodes/namespace-2 extension_point:Filter]} {Data:map[Average:NaN Perc50:NaN Perc90:NaN Perc95:NaN Perc99:NaN] Unit:ms Labels:map[Metric:scheduler_framework_extension_point_duration_seconds Name:BenchmarkPerfScheduling/PreemptionPVs/500Nodes/namespace-2 extension_point:Score]}]}: json: unsupported value: NaN

alculquicondor commented:
/assign @sanposhiho 
for a first pass

pohly commented:
/retest

sanposhiho commented on pkg/scheduler/framework/plugins/interpodaffinity/plugin.go:
```diff
@@ -135,6 +135,6 @@ func GetNamespaceLabelsSnapshot(ns string, nsLister listersv1.NamespaceLister) (
 		return labels.Merge(podNS.Labels, nil)
 	}
-	klog.ErrorS(err, "getting namespace, assuming empty set of namespace labels", "namespace", ns)
+	klog.V(3).InfoS("getting namespace, assuming empty set of namespace labels", "namespace", ns, "err", err)
```
I think this should be error logs because the namespace of Pod should be gotten successfully. 
And, either way, we can avoid this log by creating a namespace for a Pod in scheduler_perf. 
I created PR for this: #107700
@alculquicondor  Please take a look at this 🙇

pohly replied:
If it's an error, then the function should return an error and whoever calls it needs to deal with it (abort the operation or log the error and proceed, depending on what's appropriate).

sanposhiho replied:
Yeah, I also like the way you are saying; return error from this kind of utility func and log error (or abort scheduling) in who calls it. 
I still think it should be error log, but it is not a strong opinion. Anyway, it is not big deal here cuz #107700 will resolve this error log.

pohly replied:
I briefly looked at what it take to return the error, but it gets called in a few places and I don't know any of them, so I'd prefer to not touch them.

pohly replied:
Odd ordering of messages here :grin: 

There''s no particular reason for `V(3)`. It's defined as "Extended information about changes". `V(2)` is the recommended default ("Useful steady state information about the service and important log messages that may correlate to significant changes in the system. ").

Shall I change it to `V(2)`?

pohly replied:
/hold

alculquicondor replied:
I see. Where are those definitions?

It sounds like it fits 3 better.

pohly replied:
From https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md#logging-conventions

I've kept the level and updated the message as you suggested. Please LGTM again so that this can be merged once we have approval.

pohly replied:
/hold cancel

alculquicondor commented:
/approve

pohly commented:
I've rebased due to a conflict.

Can I get this reviewed? @alculquicondor, perhaps you can start for SIG Scheduling?

alculquicondor commented:
I already reviewed and only left 1 comment.

But lgtm anyways

/lgtm

fedebongio commented:
/remove-sig api-machinery

pohly commented:
/assign @sttts

For approval.

alculquicondor commented:
/lgtm

pohly commented:
/retest

pohly commented:
/retest

pohly commented:
/hold

pull-kubernetes-integration is failing rather consistently. I'll try to debug that tomorrow.

pohly commented:
/retest
/hold cancel

Looks like pull-kubernetes-integration is flaky independently of this PR here.

pohly commented:
/retest

fedebongio commented:
/triage accepted

pohly commented:
/retest

oomichi commented:
`./test` part seems good for me.

/lgtm
/approve

sttts commented on pkg/controlplane/controller.go:
```diff
@@ -163,9 +163,7 @@ func (c *Controller) Start() {
-	if err := c.EndpointReconciler.RemoveEndpoints(kubernetesServiceName, c.PublicIP, endpointPorts); err == nil {
-		klog.Error("Found stale data, removed previous endpoints on kubernetes service, apiserver didn't exit successfully previously")
-	} else if !storage.IsNotFound(err) {
+	if err := c.EndpointReconciler.RemoveEndpoints(kubernetesServiceName, c.PublicIP, endpointPorts); err != nil && !storage.IsNotFound(err) {
```
@aojea wasn't that your code?

pohly replied:
See the commit message: https://github.com/kubernetes/kubernetes/pull/107677/commits/be4bf5cb20c18fa9b74e5aff7fc64639b1797bd3

pohly replied:
@aojea : does that explain it?

aojea replied:
heh, I understand now, this thing bit me twice this week
The test is using the masterCount reconciler, not the lease reconciler. The lease reconciler is the default ... 
The integration framework should use the lease reconciler ... I think I will write something to try to deprecate the masterCount reconciler, it is flaky :(

aojea replied:
https://github.com/kubernetes/kubernetes/blob/76cdb57ccfbfebc689fbce45f289add8a0562e07/cmd/kube-apiserver/app/options/options.go#L114

aojea replied:
I think we should keep the code and deprecate the master-count
https://github.com/kubernetes/kubernetes/pull/107952

pohly replied:
As @aojea is going to address the false error message, I took this commit out of this PR.

pohly replied:
The PR was closed without a conclusion, but the issue about the error message is still open: https://github.com/kubernetes/kubernetes/issues/107953

sttts commented on test/integration/framework/controlplane_utils.go:
```diff
@@ -65,6 +65,13 @@ const (
+	// line. k8s.io/component-base/logs called klog.InitFlags for us.
+	v := flag.Lookup("v").Value
+	v.Set("4")
+}
```
why? This is affecting everybody who imports. Before it was just effective if you created an apiserver.

pohly replied:
The "why" is that the previous approach didn't work for -v0 on the command line, which is relevant for benchmarks.

I was assuming that setting -v4 as default would be desirable for all integration tests. If that is not the case, then `startAPIServerOrDie` must allow controlling whether it raises the verbosity.

Would `controlplane.Config` be a suitable place for that?

pohly replied:
> Would controlplane.Config be a suitable place for that?

Hmm, probably not.

pohly replied:
Simpler than an explicit parameter would be a global variable in controlplane_utils.go:
```
// MinVerbosity determines the minimum klog verbosity when running tests that involve the apiserver.
// This overrides the -v value from the command line, i.e. -v0 has no effect when MinVerbosity is 4 (the default).
// Tests can opt out of this by setting MinVerbosity before starting the control plane.
var MinVerbosity = 4
```

@dims: I think you added that code originally. Any thoughts?

pohly replied:
I replaced the init function with the global variable.

@mtaufen : can you take another look?

pohly committed:
scheduler_perf: run with -v=0 by default

This provides a mechanism for overriding the forced increase of the klog
verbosity to 4 when starting the apiserver and uses that for the scheduler_perf
benchmark. Other tests run as before.

A global variable was used because adding an explicit parameter to several
helper functions would have caused a lot of code churn (test ->
integration/util.StartApiserver ->
integration/framework.RunAnAPIServerUsingServer ->
integration/framework.startAPIServerOrDie).

pohly commented:
@sttts : can you take another look?

sttts commented:
Change looks good now from apiserver point of view.

@alculquicondor @oomichi can you re-lgtm?

alculquicondor commented on test/integration/scheduler_perf/util.go:
```diff
@@ -75,6 +76,8 @@ func newDefaultComponentConfig() (*config.KubeSchedulerConfiguration, error) {
 //   - client rate limit is set to 5000.
 func mustSetupScheduler(config *config.KubeSchedulerConfiguration) (util.ShutdownFunc, coreinformers.PodInformer, clientset.Interface, dynamic.Interface) {
+	// Run API server with minimimal logging.
+	framework.MinVerbosity = 0
```
So we cannot longer control this via a command line flag?

pohly replied:
This is a minimum, not the final value. In `test/integration/scheduler_perf` all values for `-v` work as expected, i.e. `-v=1` runs at verbosity 1.

It's everywhere else that `-v=1` gets bumped up to 4 when the apiserver is started. This seemed odd to me, but apparently was desired.

alculquicondor commented:
/lgtm
